### PR TITLE
Personal/sowu/readme generalization

### PR DESCRIPTION
**Pull request by sowu880**

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -436,6 +490,23 @@ There are a few parameters that can help you customize the noise amount for diff
 
 Note that the target field should be of either a numeric type (integer, decimal, unsignedInt, positiveInt) or a quantity type (Quantity, SimpleQuantity, Money, etc.). 
 
+## Generalize method
+As one of the anonymization method, generalization means mapping values to the higher level of generalization. It is the process of abstracting distinguishing value into a more general, less distinguishing value. Generalization attempts to preserve data utility while also reducing the identifiability of the data. 
+Generalization using FHIRPath predicate expression to define a set of cases that specify the condition and target value like [sample rules](#Sample-rules-using-FHIRPath). Follows are some examples of cases.
+
+|Data Type|Expression|Explain|Input data-> Output data|
```
Comment (BoyaWu10) [inline README.md:497]: Explanation?

File: README.md [hunk 2]
```diff
@@ -436,6 +490,23 @@ There are a few parameters that can help you customize the noise amount for diff
...
+|-----|-----|-----|-----|
+|numeric|_"$this>=0 and $this<20": "20"_|Data fall in the range [0,20) will be replaced with 20. |18 -> 20|
+|numeric|_"true": "($this div 10)*10"_|Approximate data to multiples of 10. |18 -> 10|
+|string| _"$this in ('es-AR' \| 'es-ES' \| 'es-UY')": "'es'"_|Data fall in the value set will be mapped to "es".|'es-UY' -> 'es'|
+|string| _"$this.startsWith(\'123\')": "$this.subString(0,2)+\'*\*\*\*\' "_ |Mask sensitive string code.|'1230005' -> '123****'|
+|date, dateTime, time|_"$this >= @2010-1-1": "@2010"_|Data fall in a date/time/dateTime range is mapped to one date/time/dateTime value.| 2016-03-10 -> 2010|
+|date, dateTime, time|_"$this.replaceMatches('(?&lt;year&gt;\\\d{2,4})-(?&lt;month&gt;\\\d{1,2})-(?&lt;day&gt;\\\d{1,2})\\\b', '${year}-${month}'"_|Omit "day" to generalize specific time.|2016-01-01 -> 2016-01|
+
+For each generalization rule, there are several additional settings to specify in configuration files:
+- [required] **cases** An object defines key-value pairs to define case condition and replacement value using FHIRPath predicate expression. key is the condition and value is the target value.
```
[the remark this answers is not in the document]
Comment (sowu880) [reply README.md:507]: "key" in key-value pair. Is there any better expression?

File: README.md [hunk 3]
```diff
@@ -492,21 +492,21 @@ Note that the target field should be of either a numeric type (integer, decimal,
...
 |numeric|_"true": "($this div 10)*10"_|Approximate data to multiples of 10. |18 -> 10|
 |string| _"$this in ('es-AR' \| 'es-ES' \| 'es-UY')": "'es'"_|Data fall in the value set will be mapped to "es".|'es-UY' -> 'es'|
 |string| _"$this.startsWith(\'123\')": "$this.subString(0,2)+\'*\*\*\*\' "_ |Mask sensitive string code.|'1230005' -> '123****'|
-|date, dateTime, time|_"$this >= @2010-1-1": "@2010"_|Data fall in a date/time/dateTime range is mapped to one date/time/dateTime value.| 2016-03-10 -> 2010|
-|date, dateTime, time|_"$this.replaceMatches('(?&lt;year&gt;\\\d{2,4})-(?&lt;month&gt;\\\d{1,2})-(?&lt;day&gt;\\\d{1,2})\\\b', '${year}-${month}'"_|Omit "day" to generalize specific time.|2016-01-01 -> 2016-01|
+|date, dateTime, time|_"$this >= @2010-1-1": "@2010"_|Data fall in a date/time/dateTime range will be mapped to one date/time/dateTime value.| 2016-03-10 -> 2010|
+|date, dateTime, time|_"$this.replaceMatches('(?&lt;year&gt;\\\d{2,4})-(?&lt;month&gt;\\\d{1,2})-(?&lt;day&gt;\\\d{1,2})\\\b', '${year}-${month}'"_|Omit "day" to generalize specific data.|2016-01-01 -> 2016-01|
```
Comment (BoyaWu10) [inline README.md:504]: date or data?

**Review by skywalker-wu — APPROVED** [review]

File: README.md [hunk 4]
```diff
@@ -296,7 +296,8 @@ The elements can be specified using [FHIRPath](http://hl7.org/fhirpath/) syntax.
 |perturb|Elements of numeric and quantity types| [Perturb](#Perturb-method) the value with random noise addition.  |
 |cryptoHash|All elements| Transforms the value using [Crypto-hash method](#Crypto-hash-method). |
 |encrypt|All elements| Transforms the value using [Encrypt method](#Encrypt-method).  |
-|substitute|All elements| [Substitutes](#Substitute-method) the value to a predefined value.  |
+|substitute|All elements| [Substitutes](#Substitute-method) the value to a predefined value. |
+|generalize|Elements of primitive types|[Generalize](#Generalize-method) the value into a more general, less distinguishing value.
```
Comment (ranvijaykumar) [inline README.md:300]: Generalize -> Generalizes

File: README.md [hunk 5]
```diff
@@ -436,6 +490,25 @@ There are a few parameters that can help you customize the noise amount for diff
 
 Note that the target field should be of either a numeric type (integer, decimal, unsignedInt, positiveInt) or a quantity type (Quantity, SimpleQuantity, Money, etc.). 
 
+## Generalize method
+As one of the anonymization method, generalization means mapping values to the higher level of generalization. It is the process of abstracting distinguishing value into a more general, less distinguishing value. Generalization attempts to preserve data utility while also reducing the identifiability of the data.
```
Comment (ranvijaykumar) [inline README.md:494]: As one of the anonymization method -> As one of the anonymization methods

File: README.md [hunk 6]
```diff
@@ -436,6 +490,25 @@ There are a few parameters that can help you customize the noise amount for diff
...
+|-----|-----|-----|-----|
+|numeric|_"$this>=0 and $this<20": "20"_|Data fall in the range [0,20) will be replaced with 20. |18 -> 20|
+|numeric|_"true": "($this div 10)*10"_|Approximate data to multiples of 10. |18 -> 10|
+|string| _"$this in ('es-AR' \| 'es-ES' \| 'es-UY')": "'es'"_|Data fall in the value set will be mapped to "es".|'es-UY' -> 'es'|
+|string| _"$this.startsWith(\'123\')": "$this.subString(0,2)+\'*\*\*\*\' "_ |Mask sensitive string code.|'1230005' -> '123****'|
+|date, dateTime, time|_"$this >= @2010-1-1": "@2010"_|Data fall in a date/time/dateTime range will be mapped to one date/time/dateTime value.| 2016-03-10 -> 2010|
+|date, dateTime, time|_"$this.replaceMatches('(?&lt;year&gt;\\\d{2,4})-(?&lt;month&gt;\\\d{1,2})-(?&lt;day&gt;\\\d{1,2})\\\b', '${year}-${month}'"_|Omit "day" to generalize specific date.|2016-01-01 -> 2016-01|
+
+For each generalization rule, there are several additional settings to specify in configuration files:
+- [required] **cases** An object defines key-value pairs to define case condition and replacement value using FHIRPath predicate expression. _key_ represents case condition and _value_ represents target value.
```
Comment (ranvijaykumar) [inline README.md:507]: An object defines key-value pairs to define -> An object defining key-value pairs to specify

File: README.md [hunk 7]
```diff
@@ -436,6 +490,25 @@ There are a few parameters that can help you customize the noise amount for diff
...
+
+- [optional] **otherValues** Define the operation for values mismatch all cases. The value could be "redact" or "keep". The default value is "redact".
```
Comment (ranvijaykumar) [inline README.md:509]: Define the operation for values mismatch all cases -> Define the operation for values that do not match any of the cases.

**Review by skywalker-wu — APPROVED** [review]

**Review by moria97 — APPROVED** [review]

**Pull request merged**